app/models/term.rb

```diff
@@ -0,0 +1,105 @@
+class Term < ApplicationRecord
+  TEMPORARY_URI_BASE = 'temp:'.freeze
+  LOCAL     = 'local'.freeze
+  TEMPORARY = 'temporary'.freeze
+  EXTERNAL  = 'external'.freeze
+
+  TERM_TYPES = [LOCAL, EXTERNAL, TEMPORARY].freeze
+
+  VALID_URI_REGEX = /\A#{URI.regexp}\z/
+
+  belongs_to :vocabulary
+
+  before_validation :add_uuid, :add_uri, :add_uri_hash, on: :create
+
+  after_commit :update_solr # Is triggered after successful save/update/destroy.
+
+  validates :vocabulary, :pref_label, :uri, :uri_hash, :uuid, :term_type, presence: true
+  validates :term_type, inclusion: { in: TERM_TYPES }
+  validate  :uuid_uri_and_term_type_unchanged
+
+  store :custom_fields, coder: JSON
+
+  serialize :alt_label, Array
+
+  def to_solr
+    {
+      'uuid'          => uuid,
+      'uri'           => uri,
+      'pref_label'    => pref_label,
+      'alt_label'     => alt_label,
+      'term_type'     => term_type,
+      'vocabulary'    => vocabulary.string_key,
+      'authority'     => authority,
+      'custom_fields' => custom_fields.to_json
+    }.tap do |doc|
+      vocabulary.custom_fields.each do |k, v|
+        doc["#{k}#{URIService.solr_suffix(v[:data_type])}"] = self.custom_fields[k]
+      end
+    end
+  end
+
+  def add_custom_field(field, value)
+    if vocabulary.nil?
+      errors.add(:custom_field, 'cannot add custom field until vocabulary relationship has been set')
+    elsif !vocabulary.custom_fields.keys.include?(field)
+      errors.add(:custom_field, "cannot add #{field} because it is not defined by parent vocabulary")
+    else
+      self.custom_fields[field] = value
+    end
+  end
+
+  private
+
+    def add_uuid
+      self.uuid = SecureRandom.uuid unless uuid
```

Comment:
Should this raise an error instead of silently failing?  If anything is calling add_uuid twice for a term, that would be a problem.

Reply:
I am a little confused by your comment. I am checking whether or not a uuid is set because I am allowing for a uuid to be set at creation. That way when we migrate the uuid can stay the same for local terms.

Reply:
We decided that this method will only set the uuid for new records, if the record has already been persisted an error will be raised. UUID validation should also be added.